### PR TITLE
Fix DDP synchronization: Add window broadcasting and sequential sampling mode

### PR DESCRIPTION
**Pull request by azadeh-gh**
Closes #[#73 ]

## Problem
Previously, each DDP rank independently selected different time windows for training, causing gradient averaging across incompatible data distributions. This violated the fundamental assumption of DDP that all ranks should compute gradients from the same distribution.

See #73  for detailed mathematical explanation.

### 1. DDP Window Synchronization ([callbacks.py](https://github.com/NOAA-EMC/ocelot/blob/feature/sequential-sampling-fixes/gnn_model/callbacks.py))
- Added `_broadcast_window()` method to synchronize window selection across all ranks
- Rank-0 selects the window, broadcasts to all other ranks via `dist.broadcast_object_list()`
- Ensures all ranks use identical `(start_date, end_date)` each epoch
- Applies to both `ResampleDataCallback` (random mode) and `SequentialDataCallback` (sequential mode)

### 2. Sequential Sampling Mode ([gnn_datamodule.py](https://github.com/NOAA-EMC/ocelot/blob/feature/sequential-sampling-fixes/gnn_model/gnn_datamodule.py))
- **New `BalancedSequentialShard` sampler**: Distributes bins in chronological order across ranks
  - Each rank gets a contiguous chunk of bins (no shuffling)
  - Maintains temporal continuity for sequential learning
- **Dynamic worker configuration**:
  - Random mode: 4 workers (parallel loading, order doesn't matter)
  - Sequential mode: 1 worker (preserves chronological order)
  - Validation: Always 1 worker (needed for FSOI)

### 3. Training Script Updates ([train_gnn.py](https://github.com/NOAA-EMC/ocelot/blob/feature/sequential-sampling-fixes/gnn_model/train_gnn.py))
- Added `--sampling_mode` argument: `random` (default) or `sequential`
- Added `--window_mode` argument: `random` (default) or `sequential`
- Updated callback selection logic based on modes

### Random Mode (default)
```bash
python train_gnn.py --sampling_mode random --window_mode random

### CODE REVIEW

**Review by xincjin-NOAA — DISMISSED** [review]
looks good to me.
